In [1]:
import numpy as np
np.__version__

'1.26.4'

In [2]:
import sklearn
from sklearn.datasets import load_iris
sklearn.__version__

'1.5.2'

In this module we will use the `sklearn` library to load a real dataset.

You can find the documentation here: https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris

In [3]:
iris_data = load_iris()

In [4]:
iris_data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

This dataset contains measurements of 4 flower structures from 3 plant species.

In this exercise, we will only use the measurements of the 4 structures.

In [5]:
iris_features = iris_data['data']

In [6]:
iris_features

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [7]:
iris_features.shape

(150, 4)

👆 There are 150 flowers, and 4 measurements.

#### `np.percentile`

Calculate the 10th, 25th and 50th percentiles of the 4 variables, that is, "along the rows".

In [8]:
np.percentile(a=iris_features, q=10, axis=0)

array([4.8, 2.5, 1.4, 0.2])

👆 This means that 10% of plants have the first measurement/variable less than or equal to 4.8, the second less than or equal to 2.5, the third less than or equal to 1.4 and the fourth less than or equal to 0.2.

In [9]:
np.percentile(a=iris_features, q=25, axis=0)

array([5.1, 2.8, 1.6, 0.3])

👆 This means that 25% (1/4) of the plants have the first measurement less than or equal to 5.1, the second less than or equal to 2.8, the third less than or equal to 1.6 and the fourth less than or equal to 0.3.

In [10]:
np.percentile(a=iris_features, q=50, axis=0)

array([5.8 , 3.  , 4.35, 1.3 ])

🤔 Can you interpret this result?

#### `np.quantile`

Repeat the previous exercise but now using the `np.quantile` function. Use the same values ​​for the quantiles, but remember that quantiles are percentiles divided by 100. Note that the results are the same, as expected.

In [11]:
np.quantile(a=iris_features, q=0.1, axis=0)

array([4.8, 2.5, 1.4, 0.2])

In [12]:
np.quantile(a=iris_features, q=0.25, axis=0)

array([5.1, 2.8, 1.6, 0.3])

In [13]:
np.quantile(a=iris_features, q=0.5, axis=0)

array([5.8 , 3.  , 4.35, 1.3 ])

#### `np.median`, `np.mean`

Calculate the median and mean of the 4 variables.

In [14]:
np.median(a=iris_features, axis=0)

array([5.8 , 3.  , 4.35, 1.3 ])

👆 The median is equal to the 50th percentile and the 0.5 quantile.

In [15]:
np.mean(a=iris_features, axis=0)

array([5.84333333, 3.05733333, 3.758     , 1.19933333])

#### `np.std`, `np.var`

Calculate the standard deviation and variance of the 4 variables.

In [16]:
iris_std = np.std(a=iris_features, axis=0)

In [17]:
iris_std

array([0.82530129, 0.43441097, 1.75940407, 0.75969263])

In [18]:
iris_var = np.var(a=iris_features, axis=0)

In [19]:
iris_var

array([0.68112222, 0.18871289, 3.09550267, 0.57713289])

Use `np.allclose` to confirm that the variance is equal to the squared standard deviation (`np.power`).

In [20]:
np.allclose(np.power(iris_std, 2), iris_var)

True

#### `np.cov`

Determine the covariance between the 4 variables. Tip: you will need to use the array in its transposed form.

In [21]:
np.cov(m=iris_features.T)

array([[ 0.68569351, -0.042434  ,  1.27431544,  0.51627069],
       [-0.042434  ,  0.18997942, -0.32965638, -0.12163937],
       [ 1.27431544, -0.32965638,  3.11627785,  1.2956094 ],
       [ 0.51627069, -0.12163937,  1.2956094 ,  0.58100626]])

#### `np.corrcoef`

Determine the Pearson correlation coefficient between the 4 variables. Again, you will need to use the transposed form of the array.

In [22]:
np.corrcoef(x=iris_features.T)

array([[ 1.        , -0.11756978,  0.87175378,  0.81794113],
       [-0.11756978,  1.        , -0.4284401 , -0.36612593],
       [ 0.87175378, -0.4284401 ,  1.        ,  0.96286543],
       [ 0.81794113, -0.36612593,  0.96286543,  1.        ]])

#### `np.histogram`

Determine the value count and intervals to construct a histogram of the variable indexed 2, with 5 intervals in total.

In [23]:
hist, bin_edges = np.histogram(a=iris_features[:, 2], bins=5)

In [24]:
hist, bin_edges

(array([50,  3, 34, 47, 16]), array([1.  , 2.18, 3.36, 4.54, 5.72, 6.9 ]))

👆 Note that the majority of plants (50 in total) are in the lower range of the intervals, with values ​​between 1 and 2.18.

#### `np.bincount`

Use `np.bincount` to determine the frequency of each flower in unit intervals, in relation to the variable indexed 2. You will need to enter the values ​​as integers (`astype` method).

In [25]:
np.bincount(iris_features[:, 2].astype(int))

array([ 0, 50,  0, 11, 43, 35, 11])

👆 There are no flowers in range 0 to 1, but 50 flowers in range 1 to 2.